### ISSUE TITLE
Feature: getRotation in MapObject (Tiled Editor)

### ISSUE THREAD

**Issue by Nodragem** [libgdx/libgdx#2742]
Hi,

I notice that libGDX MapObject (used to interface with Tiled Map Object, see the [Tiled Map Editor] (http://www.mapeditor.org/) ) does not have a method `getRotation()`. I also tried something like: `myMapObject.getProperties().get("rotation", String.class)`.

I think it could be easy to add this feature?

For now, I import the rotation in libGDX by copying by hand the _rotation property_ into a _custom property_ named `"myRotation"` (in Tiled).

However, I noticed that libGDX has the position origin of its Objects on the bottom-left corner, while Tiled works with a position origin on the top-left corner.
For now, libGDX do the position translation by removing 1 on the y-axis.

This basic translation won't work if we want to take in account the rotation of Tiled objects.
If we want to correct it, for now, we need to add this vector to our MapObject position:
``` java
float rotation = Float.parseFloat(myMapObj.getProperties().get("myRotation", String.class));
float ysize = myMapObj.getsize().y

Vector2 correctionPos = new Vector2(0 + ysize*MathUtils.cosDeg(-rotation-90),
                                    ysize + ysize*MathUtils.sinDeg(-rotation-90));

myMapObject.getPosition(new Vector2()).add(correctionPos)
```

The +1 in the y-axis of `correctionPos` is to remove the basic translation made by libGDX.
The Cos and Sin of the complementary angle is the general solution to change from a top-left origin system to a bottom-left origin system (Note you need to multiply it by the ysize of your rectangle.). 
Note also that a rotation `r` in Tiled Editor is a rotation of `-r` is libGDX.

So I hope I bring with me every thing to allow an implemetation of the feature :D maybe I should try to do a pull request, but I never tried.

PS: sorry if the feature already exists, I'm maybe on a old version of libGDX :/

EDIT:
I corrected the `correctionPos` vector. I replaced `1 + ysize*MathUtils.sinDeg(-rotation-90))` with `ysize + ysize*MathUtils.sinDeg(-rotation-90))` .

**Comment by MobiDevelop** [comment]
Rotation is added by the loader to the MapObject properties as a Float, if it exists so you should not have to populate an additional custom property.

**Comment by Nodragem** [comment]
That's right, however, when the rotation equals 0 in the Tiled Editor, the MapObject loader return `null` for the rotation. (Note that the rotation property _always_ exist in the Tiled Editor)

To be clearer, instead of doing that:
``` java
float rotation = rectangleObj.getProperties().get("rotation", Float.class);
```

We have to do that :
``` java
float rotation =0;
if (rectangleObj.getProperties().get("rotation", Float.class) != null)
         rotation = rectangleObj.getProperties().get("rotation", Float.class);
```

Which may not be intuitive and practical?

__Second point:__
So, OK, we can get the rotation, but what about my story of positionning?
It still seems to me that libGDX gives a wrong position for rotated object (see the correction I suggest above).

I should specify I am on libGDX 1.4.1.

**Comment by MobiDevelop** [comment]
We don't do anything with the `rotation` within libgdx, so if something is not rotating correctly, that is something that needs to be solved when you are using the value. We simply pass the value along. As for the value always being there in Tiled, it must not be there in the tmx file, otherwise you would get a rotation value in the properties.

**Comment by Nodragem** [comment]
Ok, I will report an issue in the Tiled Editor project concerning the absence of the propertie `rotation` in the tmx file when the rotation is at zero in the editor (I just checked, you're right :D ). But, maybe libGDX could conclude by itself that if the rotation property of a MapObject is not here, it is equal 0?

Aside, there is still two problems:
 1. As you say, you _simply_ return the value `rotation` given by Tiled. Tiled has its y-axis going down while libGDX has its y-axis going up, then we need to get the `-rotation` to get the good rotation.

 2. The positionning problem. LibGDX codes the position of its objects with the bottom-left corner, while Tiled codes the position of its objects with the top-left corner.
For now, the conversion in libGDX from one system to the other is done by removing the y-size of objects from their y-axis coordinate. This method works well when there is no rotation.

So basically, here what is happening if we just take the rotation given back by the MapObject loader:
![example_forum](https://cloud.githubusercontent.com/assets/10520249/5745923/eae9c93e-9c23-11e4-9914-5a4337bb58b5.png)

LibGDX surely applies the following vector to correct Tiled position system:
``` java
Vector2 correctionPosition = new Vector2(0, ysize);
```
Me, I suggest that libGDX uses this corrective vector instead:
``` java
Vector2 correctionPosition = new Vector2(ysize*MathUtils.cosDeg(-rotation-90),
                                    ysize*MathUtils.sinDeg(-rotation-90));
```
For a rotation = 0, it is equivalent to the precedent correction.

**Comment by MobiDevelop** [comment]
I get what you are saying, however, how you use a property is really up to you.

**Comment by Nodragem** [comment]
I don't think it should be to the user to correct the MapObject positions. 

Especially when libGDX is _already_ doing a basic correction by removing the y-size on the y-axis coordinate. This means libGDX is _already_ assuming that the position from the Mapping software are given according to the top-left corner position of objects. 

So instead of doing a basic correction, why not doing the right correction which supports rotation? Especially when it is easy to change?

To simplify the life of libGDX users is a good thing to do :) 
When we do a rotation in Tiled, it should be usable directly in libGBX.

**Comment by Nodragem** [comment]
I still think that the position/rotation correction should be integrated in libGDX.
I mean, this correction will be transparent to people who don't use the rotation property in Tiled, and it will make life easier for people who do use the rotation property in Tiled.
So it is a win-win situation ^^

That's actually weird that nobody had the same problem :/ ... ?
Anyway, I would like the opinion of other people than MobiDevelop. Like that we can decide to close this issue thread or decide to work on it.

Have all a good day!

**Comment by badlogic** [comment]
You have a point in that we correct for y-up in most cases. I'm undecided whether we should correct rotation as well. @MobiDevelop your call, please close this issue after your decision :)

**Comment by MobiDevelop** [comment]
We'll not modify the position to account for rotation (Tiled also doesn't do this). If you'd like, we can invert the rotation degrees if that would be helpful.

**Comment by Nodragem** [comment]
Yes, already to invert the rotation angle is a good thing (that's the y-up correction for the rotation).

Your argument 
>"Tiled also doesn't do this"

is a no sense unfortunately :/ Tiled doesn't need to correct its own position as its rotation is done on the same axis (y heading down, x heading right).

I just feel like you should either do the y-up correction properly or not at all.
1) To just add +height to the y-position is a dirty/easy y-up correction which only works with rotation = 0.
2) To compute the cos and sinus is the general/complicated y-up correction which works with all rotation.

I don't even understand why there is a discussion about integrating the general correction ^^ what are the drawbacks? What is so detrimental to libGDX to refuse so strongly this enhancement?

I think that one good thing of libGDX is to easily work with Tiled. So, in my opinion, any improvements in that sense should be done.

__To do an aside:__
I have the impression that libGDX developpers are stubborn on the fact that libGDX has to stay a Framework and not become an engine. 
However, knowing that nobody actually has a clear definition of Engine and Framework, the trend is to slow down improvements to libGDX, by fear that any improvement could make libGDX goign close to an Engine ^^.

It is a paradoxal idea that a Framework has to let everything to be done by the user.
If the user was using a Framework to do everything himself, he would not use the framework at all, isn't?

The difference between an Engine and a Framework, in my opinion, is NOT that the user is more helped or less helped. 
The difference, in my opinion, is that a Engine is genre-specific ( genre as  platformer engine, FPS engine, puzzle engine), while the Framework is genre-non-specific.

To be non-specific, there is two way: either you don't do anything, either you try to make easy anything which is shared by the majority of genres.
Here the compatibility with Tiled can be seen as something important for many kinds of 2d games.

Maybe libGDX roadmap and philosophy needs to be defined clearly. In order to have a reference when we accept/reject an enhancement.

__End of the aside__

**Comment by MobiDevelop** [comment]
Tiled does not adjust the position of the object when you rotate it. It rotates about its origin - you simply have to change the origin for your rotation to the top-left corner. That is dead simple to do with an renderer provided by libgdx.

**Comment by Nodragem** [comment]
That's indeed an other way to make it :)

```
float rotation =0;
if (rectangleObj.getProperties().get("rotation", Float.class) != null)
         rotation = rectangleObj.getProperties().get("rotation", Float.class);

float ysize = myMapObj.getsize().y
Vector2 tiledpositiion = myMapObject.getPosition(new Vector2());
Vector2 tiledorigin = new Vector2(0, ysize);
```

Then you use the tiled origin in your renderer. Perfect.

What about the bodies of Box2d? is that dead simple solution stays simple to implement? 
I didn't find how to define a centre of rotation with box2d :/
Feel free to post the solution for future reference.
